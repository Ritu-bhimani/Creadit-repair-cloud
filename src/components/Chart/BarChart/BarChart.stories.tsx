import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ChartStyles } from '../styles';
import { BarChartComponent } from './BarChart';

export default {
    title: 'Bar chart',
    component: BarChartComponent
} as ComponentMeta<typeof BarChartComponent>;

const Template: ComponentStory<typeof BarChartComponent> = (args) => (
    <BarChartComponent {...args} />
);
let tooltipType = '';
const selectedToolTip = (val: any | undefined) => {
    if (val !== null)
        tooltipType = val
};
type CustomToolTipProps = {
    active?: boolean;
    payload?: any;
    label?: string;
    cursor?: any;
    style?: any;
    primary?: boolean;
}
const CustomTooltip = (props: CustomToolTipProps) => {
    if (props.active && props.payload && props.payload.length && props.primary) {
        return (
            <div css={ChartStyles.customTooltip} >
                <p className="label">{`${props.payload[0].value} Affiliates`}</p>
            </div>
        );
    } else if (props.active && props.payload && props.payload.length && !props.primary) {
        for (const bar of props.payload)
            if (bar.dataKey === tooltipType)
                return (
                    <div css={ChartStyles.customTooltip} >
                        <p className="label">{`${bar.name === 'client' ? "Active Client" : "Leads"}, ${props.label}, ${bar.value}`}</p>
                    </div>
                );
        return null
    }

    return null;
};
type renderCustomBarLabelProps = {
    payload?: any;
    x: number;
    y: number;
    width: number;
    value: number;
}
const renderCustomBarLabel = (props: renderCustomBarLabelProps) => {
    const { payload, x, y, width, value } = props
    return (
        <text
            x={x + width / 2}
            y={y}
            fill="#666"
            textAnchor="middle"
            dy={-6}
        >{`${value}`}</text>
    );
};
export const Primary = Template.bind({});

Primary.args = {
    primary: true,
    data: [
        {
            name: 'Aug 22',
            value: 2400,
        },
        {
            name: 'Sep 22',
            value: 1398,
        },
        {
            name: 'Oct 22',
            value: 9800,
        },
        {
            name: 'Nov 22',
            value: 3908,
        },
        {
            name: 'Dec 22',
            value: 3908,
        },
        {
            name: 'Jan 22',
            value: 4800,
        }
    ],
    height: 155,
    width: 500,
    customTooltip: <CustomTooltip primary={true} />,
    xAxisDataKey: "name",
    barDataKey: "value",
    renderCustomBarLabel: renderCustomBarLabel
};


export const Secondary = Template.bind({});
Secondary.args = {
    ...Primary.args,
    primary: false,
    height: 155,
    width: 500,
    customTooltip: <CustomTooltip primary={false} />,
    xAxisDataKey: "name",
    barDataKey: "lead",
    barDataKey2: "client",
    data: [
        {
            name: 'Aug 22',
            lead: 300,
            client: 2400,
        },
        {
            name: 'Sep 22',
            lead: 315,
            client: 1398,
        },
        {
            name: 'Oct 22',
            lead: 330,
            client: 9800,
        },
        {
            name: 'Nov 22',
            lead: 340,
            client: 3908,
        },
        {
            name: 'Dec 22',
            leads: 340,
            client: 3908,
        },
        {
            name: 'Jan 22',
            leads: 340,
            client: 4800,
        }
    ],
    getChartType: selectedToolTip
};


