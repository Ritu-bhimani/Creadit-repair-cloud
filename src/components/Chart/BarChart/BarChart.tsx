import { Container } from '@mui/material';
import { FC, useEffect, useState } from 'react';
import {
  Bar,
  BarChart,
  CartesianGrid,
  Legend,
  Tooltip,
  XAxis,
  YAxis
} from 'recharts';
import {
  NameType,
  ValueType
} from 'recharts/types/component/DefaultTooltipContent';
import { ChartStyles } from '../styles';

import { ContentType } from 'recharts/types/component/Tooltip';

type BarChartProps = {
  data?: any;
  height?: number;
  width?: number;
  primary?: boolean;
  customTooltip?: ContentType<ValueType, NameType>;
  xAxisDataKey?: string;
  barDataKey?: any;
  barDataKey2?: any;
  renderCustomBarLabel?: any;
  getChartType?: any;
  selectedDefaultLegend?: any;
};

type BarDetailsProps = {
  count: number;
  month: string;
};

export const BarChartComponent: FC<BarChartProps> = (props: BarChartProps) => {
  const {
    data,
    height,
    width,
    primary,
    customTooltip,
    xAxisDataKey,
    barDataKey,
    barDataKey2,
    renderCustomBarLabel,
    getChartType,
    selectedDefaultLegend
  } = props;
  const barDetails: BarDetailsProps[] = [];
  const [selectedLedgend, setSelectedLedgend] = useState<any>([]);

  useEffect(() => {
    selectedDefaultLegend && selectedDefaultLegend(selectedLedgend);
  }, [selectedLedgend]);

  data.map((details: { count: number; month: string }, i: number) => {
    barDetails.push({
      ...details,
      count: i > 0 ? barDetails[i - 1].count + details.count : details.count
    });
  });

  const renderLegend = (props: any) => {
    const { payload } = props;
    return (
      <div
        css={{
          display: 'flex',
          border: 'solid 1px #4a4a4a',
          width: '180px',
          justifyContent: 'space-between',
          padding: '0 5px'
        }}
      >
        {payload.map((entry: { value: string }, index: number) => (
          <div
            css={{ display: 'flex', alignItems: 'center' }}
            onClick={() =>
              selectedLedgend.includes(entry.value)
                ? setSelectedLedgend(
                    selectedLedgend.filter(
                      (defaultVal: string) => entry.value != defaultVal
                    )
                  )
                : setSelectedLedgend([...selectedLedgend, entry.value])
            }
          >
            <img
              src={
                selectedLedgend.includes(entry.value)
                  ? '/assets/images/disableChart.png'
                  : entry.value == 'lead'
                  ? '/assets/images/lead.png'
                  : '/assets/images/client.png'
              }
            />
            <li
              css={{ listStyle: 'none', color: '#555 !important' }}
              key={`item-${index}`}
              style={{
                color: selectedLedgend.includes(entry.value) ? 'grey' : 'black'
              }}
            >
              {entry.value == 'lead' ? 'Lead' : 'Active Client'}
            </li>
          </div>
        ))}
      </div>
    );
  };

  return (
    <Container css={ChartStyles.ChartCss}>
      <BarChart
        data={barDetails}
        height={height ? height : 200}
        width={width ? width : 490}
        margin={{ top: 20, right: 0, left: 0 }}
      >
        <CartesianGrid strokeDasharray="0 0" vertical={false} />
        <XAxis dataKey={xAxisDataKey} />
        <YAxis />
        <Tooltip
          content={customTooltip}
          cursor={{ fill: 'transparent' }}
          itemStyle={{ borderRadius: '0' }}
        />
        {primary ? (
          <Bar
            dataKey={barDataKey}
            barSize={50}
            fill="#599ddd"
            label={renderCustomBarLabel}
          />
        ) : (
          <>
            <Bar
              minPointSize={5}
              dataKey={barDataKey}
              barSize={40}
              fill={
                !selectedLedgend.includes('lead') ? '#599ddd' : 'transparent'
              }
              label={renderCustomBarLabel}
              onMouseOver={() => getChartType(barDataKey)}
            />

            <Bar
              dataKey={barDataKey2}
              barSize={40}
              fill={
                !selectedLedgend.includes('client') ? '#a5db89' : 'transparent'
              }
              label={renderCustomBarLabel}
              onMouseOver={() => getChartType(barDataKey2)}
            />

            <Legend content={renderLegend} />
          </>
        )}
      </BarChart>
    </Container>
  );
};

export default BarChartComponent;
