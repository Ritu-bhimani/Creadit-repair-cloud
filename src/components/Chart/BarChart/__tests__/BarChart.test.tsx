import { render } from '@testing-library/react';
import { BarChartComponent } from "../BarChart"
it('should render Bar chart and match the snapshot', () => {
    const component = render(<BarChartComponent data={[]} />);
    expect(component).toMatchSnapshot();
});