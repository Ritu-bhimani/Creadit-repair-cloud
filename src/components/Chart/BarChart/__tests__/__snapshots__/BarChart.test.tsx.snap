// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Bar chart and match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 490px; height: 200px;"
        >
          <svg
            class="recharts-surface"
            height="200"
            viewBox="0 0 490 200"
            width="490"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="150"
                  width="430"
                  x="60"
                  y="20"
                />
              </clippath>
            </defs>
            <g
              class="recharts-cartesian-grid"
            >
              <g
                class="recharts-cartesian-grid-horizontal"
              >
                <line
                  fill="none"
                  height="150"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="0 0"
                  width="430"
                  x="60"
                  x1="60"
                  x2="490"
                  y="20"
                  y1="20"
                  y2="20"
                />
                <line
                  fill="none"
                  height="150"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="0 0"
                  width="430"
                  x="60"
                  x1="60"
                  x2="490"
                  y="20"
                  y1="170"
                  y2="170"
                />
              </g>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="150"
                orientation="left"
                stroke="#666"
                width="60"
                x="0"
                x1="60"
                x2="60"
                y="20"
                y1="20"
                y2="170"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              />
            </g>
          </svg>
          <div
            class="recharts-legend-wrapper"
            style="position: absolute; width: 490px; height: auto; left: 0px; bottom: 0px;"
          >
            <div>
              <img
                src="[object Object]"
              />
              <li
                style="color: black;"
              >
                Active Client
              </li>
            </div>
            <div>
              <img
                src="[object Object]"
              />
              <li
                style="color: black;"
              >
                Active Client
              </li>
            </div>
          </div>
          <div
            class="recharts-tooltip-wrapper"
            role="dialog"
            style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <span
      aria-hidden="true"
      id="recharts_measurement_span"
      style="position: absolute; top: -20000px; left: 0px; padding: 0px; margin: 0px; white-space: pre;"
    >
      0
    </span>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    >
      <div
        class="recharts-wrapper"
        role="region"
        style="position: relative; cursor: default; width: 490px; height: 200px;"
      >
        <svg
          class="recharts-surface"
          height="200"
          viewBox="0 0 490 200"
          width="490"
        >
          <title />
          <desc />
          <defs>
            <clippath
              id="recharts1-clip"
            >
              <rect
                height="150"
                width="430"
                x="60"
                y="20"
              />
            </clippath>
          </defs>
          <g
            class="recharts-cartesian-grid"
          >
            <g
              class="recharts-cartesian-grid-horizontal"
            >
              <line
                fill="none"
                height="150"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="0 0"
                width="430"
                x="60"
                x1="60"
                x2="490"
                y="20"
                y1="20"
                y2="20"
              />
              <line
                fill="none"
                height="150"
                offset="[object Object]"
                stroke="#ccc"
                stroke-dasharray="0 0"
                width="430"
                x="60"
                x1="60"
                x2="490"
                y="20"
                y1="170"
                y2="170"
              />
            </g>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
          >
            <line
              class="recharts-cartesian-axis-line"
              fill="none"
              height="150"
              orientation="left"
              stroke="#666"
              width="60"
              x="0"
              x1="60"
              x2="60"
              y="20"
              y1="20"
              y2="170"
            />
            <g
              class="recharts-cartesian-axis-ticks"
            />
          </g>
        </svg>
        <div
          class="recharts-legend-wrapper"
          style="position: absolute; width: 490px; height: auto; left: 0px; bottom: 0px;"
        >
          <div>
            <img
              src="[object Object]"
            />
            <li
              style="color: black;"
            >
              Active Client
            </li>
          </div>
          <div>
            <img
              src="[object Object]"
            />
            <li
              style="color: black;"
            >
              Active Client
            </li>
          </div>
        </div>
        <div
          class="recharts-tooltip-wrapper"
          role="dialog"
          style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
          tabindex="-1"
        >
          <div
            class="recharts-default-tooltip"
            style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
          >
            <p
              class="recharts-tooltip-label"
              style="margin: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
